### PR TITLE
Add `@schema` decorator to mark namespaces as GraphQL schemas

### PR DESCRIPTION
**Pull request by steverice**
Using the `TypeSpec.GraphQL.@schema` decorator on a namespace indicates that the decorated namespace represents a GraphQL schema that should be generated by the GraphQL emitter.

Because this allows for multiple schemas to be specified in a TypeSpec source, our test host is reworked to provide a `GraphQLSchemaRecord` corresponding to each schema produced.

This commit does not actually implement any emitter functionality, but populates a state map that will be used by the emitter in the future.

### CODE REVIEW

**Comment by steverice** [comment]
Note to reviewers: the first commit is just some `package.json` metadata that I have proposed directly upstream: https://github.com/microsoft/typespec/pull/5076

File: packages/graphql/src/lib/schema.ts [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+  type Type,
+  getTypeName,
+  validateDecoratorUniqueOnNode,
+} from "@typespec/compiler";
+
+import { reportDiagnostic } from "../lib.js";
+import { useStateMap } from "./state-map.js";
+
+// This will set the namespace for all the decorators
+export const namespace = "TypeSpec.GraphQL";
```
Comment (steverice) [inline packages/graphql/src/lib/schema.ts:15]:
```suggestion
```
This should not be needed when using `export const $decorators`.

Comment (steverice) [reply packages/graphql/src/lib/schema.ts]: Turns out we do need this. When we import `schema.js` into `schema.tsp`, the definition must be namespaced in order to match the TypeSpec declaration.

So we have two different paths for namespacing:
1. `export const $decorators` in `tsp-index.js` sets the namespace explicitly for all decorators when used by external code
2. `export const namespace = ` in `schema.js` sets the namespace when the `$schema` decorator is imported directly.

Since these are two independent import paths, they [should not conflict](https://discord.com/channels/1247582902930116749/1250119513681301514/1306717174551023636).Turns out we do need this. When we import `schema.js` into `schema.tsp`, the definition must be namespaced in order to match the TypeSpec declaration.

So we have two different paths for namespacing:
1. `export const $decorators` in `tsp-index.js` sets the namespace explicitly for all decorators when used by external code
2. `export const namespace = ` in `schema.js` sets the namespace when the `$schema` decorator is imported directly.

Since these are two independent import paths, they [should not conflict](https://discord.com/channels/1247582902930116749/1250119513681301514/1306717174551023636).

File: packages/graphql/package.json [hunk 2]
```diff
@@ -13,15 +29,14 @@
       "default": "./dist/src/testing/index.js"
     }
   },
-  "peerDependencies": {
```
Comment (swatkatz) [inline packages/graphql/package.json:16]: This was suggested by Timothee

File: packages/graphql/package.json [hunk 3]
```diff
@@ -32,5 +47,20 @@
...
+    "lib/*.tsp",
+    "dist/**",
+    "!dist/test/**"
+  ],
+  "peerDependencies": {
+    "@typespec/compiler": "workspace:~"
+  },
+  "devDependencies": {
+    "@types/node": "~22.7.5",
+    "@typespec/compiler": "workspace:~",
```
Comment (swatkatz) [inline packages/graphql/package.json:60]: why do we need to add this to devDependencies?

File: packages/graphql/package.json [hunk 4]
```diff
@@ -32,5 +47,20 @@
     "lint": "eslint src/ test/ --report-unused-disable-directives --max-warnings=0",
     "lint:fix": "eslint . --report-unused-disable-directives --fix"
   },
-  "private": true
+  "files": [
```
Comment (swatkatz) [inline packages/graphql/package.json:50]: Why is this necessary?

File: packages/graphql/package.json [hunk 5]
```diff
@@ -13,15 +29,14 @@
       "default": "./dist/src/testing/index.js"
     }
   },
-  "peerDependencies": {
-    "@typespec/compiler": "workspace:~"
+  "engines": {
```
Comment (swatkatz) [inline packages/graphql/package.json:32]: Do we really need this in graphql/package.json?

File: packages/graphql/package.json [hunk 6]
```diff
@@ -1,10 +1,26 @@
 {
   "name": "@typespec/graphql",
   "version": "0.1.0",
+  "author": "Microsoft Corporation",
```
Comment (swatkatz) [inline packages/graphql/package.json:4]: Is this true?

File: packages/graphql/src/lib.ts [hunk 7]
```diff
@@ -99,4 +101,4 @@ export const libDef = {
 
 export const $lib = createTypeSpecLibrary(libDef);
 
-export const { reportDiagnostic, createDiagnostic } = $lib;
+export const { reportDiagnostic, createDiagnostic, createStateSymbol } = $lib;
```
Comment (swatkatz) [inline packages/graphql/src/lib.ts:104]: I think Timothee suggested that we don't need `createStateSymbol`: https://github.com/microsoft/typespec/pull/5033#discussion_r1834801763

File: packages/graphql/lib/schema.tsp [hunk 8]
```diff
@@ -0,0 +1,23 @@
+import "../dist/src/lib/schema.js";
+
+using TypeSpec.Reflection;
+
+namespace TypeSpec.GraphQL;
+
+namespace Schema {
```
Comment (swatkatz) [inline packages/graphql/lib/schema.tsp:7]: Would this namespace result in any conflicts?

File: packages/graphql/src/lib/schema.ts [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+export const $schema: DecoratorFunction = (
+  context: DecoratorContext,
+  target: Namespace,
+  options: Type,
+) => {
+  validateDecoratorUniqueOnNode(context, target, $schema);
+
+  if (options && options.kind !== "Model") {
```
Comment (swatkatz) [inline packages/graphql/src/lib/schema.ts:81]: Does this need to be a specific type of model i.e. `SchemaOptions`?

File: packages/graphql/src/testing/utils.ts [hunk 10]
```diff
@@ -0,0 +1,12 @@
+import { GraphQLSchema } from "graphql";
+
+export const EMPTY_SCHEMA = new GraphQLSchema({});
+
+export const EMPTY_SCHEMA_OUTPUT = `#graphql
```
Comment (swatkatz) [inline packages/graphql/src/testing/utils.ts:5]: It would be nice to generate this output using printSchema and GraphQL AST, rather than hardcoding as a string like we do in the prototype tests. E.g.

```
schemaWithDummyQuery = new GraphQLSchema({...})


getEmittedOutput(schemaWithDummyQuery, options):
    return printSchema(schemaWithDummyQuery)
```

This way we'll get all the spacing etc correct.

File: packages/graphql/test/schema.test.ts [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+    expect(schema?.type).toBe(TestNamespace);
+    expect(schema?.name).toBeUndefined();
+  });
+
+  it("Creates a schema with a specified name", async () => {
+    const [program, { TestNamespace }, diagnostics] = await compileAndDiagnose<{
+      TestNamespace: Namespace;
+    }>(`
+      @schema({name: "MySchema"})
```
Comment (swatkatz) [inline packages/graphql/test/schema.test.ts:27]: could we add a test when an invalid property is set?

```
@schema({foo: "bar"})
```

File: packages/graphql/test/test-host.ts [hunk 12]
```diff
@@ -41,13 +64,37 @@ export async function emitWithDiagnostics(
     },
   });
   const content = runner.fs.get(outputFile);
-  ok(content, "Expected to have found graphql output");
-  // Change this to whatever makes sense for the actual GraphQL emitter, probably a GraphQLSchemaRecord
-  return [content, diagnostics];
+  const schema = content
+    ? buildSchema(content, {
```
Comment (swatkatz) [inline packages/graphql/test/test-host.ts:68]: I wonder if we can run the emitter in test mode to just return the GraphQL AST directly without deserializing the content into the AST using `buildSchema`

File: packages/graphql/package.json (hunk 3 above)
Comment (steverice) [reply packages/graphql/package.json:60]: We cannot run tests without the compiler. The same is used in other TypeSpec packages.

File: packages/graphql/package.json (hunk 4 above)
Comment (steverice) [reply packages/graphql/package.json:50]: It's used when installing the package as a dependency: https://docs.npmjs.com/cli/v10/configuring-npm/package-json#files

File: packages/graphql/package.json (hunk 6 above)
Comment (steverice) [reply packages/graphql/package.json:4]: As [I mentioned](https://github.com/pinterest/typespec/pull/1#issuecomment-2471795761), these `package.json` updates have been proposed upstream on their own. So I am hoping to get MSFT's input on what should and shouldn't be in `package.json` as I think it's ultimately up to them.

Until that, I am mostly just copying things over from other packages (primarily `openapi3`).

File: packages/graphql/lib/schema.tsp (hunk 8 above)
Comment (steverice) [reply packages/graphql/lib/schema.tsp:7]: No, because it is itself namespaced. The full namespace is `TypeSpec.GraphQL.Schema`.

File: packages/graphql/src/testing/utils.ts (hunk 10 above)
Comment (steverice) [reply packages/graphql/src/testing/utils.ts:5]: Agreed. I have done that [in later work](https://github.com/pinterest/typespec/pull/2) but forgot to reconcile it here.

File: packages/graphql/test/test-host.ts (hunk 12 above)
Comment (steverice) [reply packages/graphql/test/test-host.ts:68]: Indeed, I wanted to do this, but the structure of the test `host` and `runner` built into `@typespec/compiler` doesn't really allow for this. So after looking into it for a while I gave up.

This approach matches what is done with the `OpenAPI` tests — the file is emitted, and then loaded and JSON parsed into the OpenAPI object for tests to use.

File: packages/graphql/src/lib/schema.ts (hunk 9 above)
Comment (steverice) [reply packages/graphql/src/lib/schema.ts:81]: Interesting question. I followed the pattern of the `@service` decorator here, but it seems that [in more recent work](https://github.com/timotheeguerin/typespec/blob/3653a3591f7401b0a2f21578061bdd01d164c43f/.chronus/changes/uri-templates-2024-6-25-9-3-39.md) there's a preference to use an object literal rather than a model for such options: 
> Decorator is also expecting an object value now instead of a model.

I'm not sure why; both appear to validate the input equivalently. But the nice thing about the object value approach is that we can annotate the type in the TypeScript implementation, so I'll go with that.

File: packages/graphql/test/schema.test.ts (hunk 11 above)
Comment (steverice) [reply packages/graphql/test/schema.test.ts:27]: Because of https://github.com/pinterest/typespec/pull/1/files/c0fff8ab630090bc0902ec48f797e0ac87683317#r1849148838, we no longer need to test this — it's inbuilt TypeSpec compiler behavior.

File: packages/graphql/src/lib.ts (hunk 7 above)
Comment (steverice) [reply packages/graphql/src/lib.ts:104]: Nice, this is a lot cleaner.

File: packages/graphql/test/test-host.ts (hunk 12 above)
Comment (swatkatz) [reply packages/graphql/test/test-host.ts:68]: Gotcha! Can we add this as a comment here and maybe Brian Terlson can chime in?

**Review by swatkatz — APPROVED** [review]
Just a small comment to add the test mode thing as a comment to the test runner.

**Comment by steverice** [comment]
This [was merged upstream](https://github.com/microsoft/typespec/pull/5159), so I'm closing out this PR.
